Variants and options for product

tedraykov opened the pull request:
# Variants and options

With this change, the product variants and options are correctly displayed on the product page.

In reaction commerce, an ***option*** is a sub-variant of a product, while in Vercel commerce it has a different meaning.
In Vercel commerce each product has a set of variants and each variant has a ubiquitous set of options that defines the variant. 
To understand how RC products can correctly be normalized to VC products we can look at an example:

```
iPhone 12 (main product)
  |
  |
  |--- Black (variant)
  |       |
  |       |--- 64GB (sub-variant) - $699
  |       |--- 128GB (sub-variant) - $799
  |       |--- 256GB (sub-variant) - $899
  |
  |
  |--- Purple (variant)
  |       |
  |       |--- 64GB (sub-variant) - $799
  |       |--- 128GB (sub-variant) - $899
  |       |--- 256GB (sub-variant) - $999
```

The structure of such product in RC would be:

```json
{
    "productId": "cmVhY3Rpb24vcHJvZHVjdDpRbVI1UVh5UDZiOEpYcHNqbQ==",
    "title": "iPhone 12",
    "variants": [
        {
            "variantId": "black-iphone"
            "attributeLabel": "Color",
            "optionTitle": "Black",
            "options": [
                {
                    "variantId": "black-iphone-64gb"    
                    "attributeLabel": "Size"
                    "optionTitle": "64GB"
                },
                {
                    "variantId": "black-iphone-128gb"  
                    "attributeLabel": "Size"
                    "optionTitle": "128GB"
                }
                ...
            ]
        },
        {
            "attributeLabel": "Color",
            "optionTitle": "Purple",
            "options": [
                {
                    "variantId": "purple-iphone-64gb"
                    "attributeLabel": "Size"
                    "optionTitle": "64GB"
                },
                {    
                    "variantId": "purple-iphone-128gb"
                    "attributeLabel": "Size"
                    "optionTitle": "128GB"
                }
                ...
            ]        
        }
    ]
    ...
}
```

In Vercel Commerce terms, we have a product called "iPhone 12" that have 4 variants:

- iPhone 12 - Black 64GB
  
- iPhone 12 - Black 128GB
  
- iPhone 12 - Purple 64GB
  
- iPhone 12 - Purple 128GB
  

Each variant is uniquely defined by two options - color and size(capacity)

We can normalize a RC product to a VC one by flattening each RC option into a VC variant and setting its options to the RC option "option" and the parent RC variant "option".

```json

// RC product variant
"variantId": "black-iphone"
"attributeLabel": "Color",
"optionTitle": "Black",
"options": [
    {
        "variantId": "black-iphone-64gb"    
        "attributeLabel": "Size"
        "optionTitle": "64GB"
    }
]

// VC product variant
"variantId": "black-iphone-64gb",
"options": [
    {
        "displayName": "Color",
        "values": [{"label": "orange"}]
    },
    {
        "displayName": "Size",
        "values": [{"label": "64GB"}]
    }     
]

```

After we have defined the variant structure we have to set all possible options that a product can have. In order to to this, we have to itterate all variants and options of a RC product and get all unique options. Then we group them by `attributeLabel`

In the iPhone example the options look like:

```json
options: Array(4)
0: {
    __typename: "MultipleChoiceOption",
    id: "cmVhY3Rpb24vY2F0YWxvZ1Byb2R1Y3RWYXJpYW50Ok52U1lwcjdDdUppRHg5d1dG",
    displayName: "Size",
    values: ["64GB", "128GB", "256"]
    }
1: {
    __typename: "MultipleChoiceOption",
    id: "cmVhY3Rpb24vY2F0YWxvZ1Byb2R1Y3RWYXJpYW50Okg4aUJrTW5xa1JTckY5NTli",
    displayName: "Color",
    values: ["Black", "Purple"]
    }
length: 2
```

Now, when we add an item to the cart, Vercel Commerce tries to find a variant that contains all of the selected options.

tedraykov committed:
Implement product attributes (variants)

tedraykov committed:
Graphql codegen change

When generating graphql types, the generation of optional types is
redundant because the value is wrapped in Maybe type anyway.
Removing the redundancy simplifies the type checking whenever generated
types are used.

tedraykov commented on framework/reactioncommerce/utils/normalize.ts:
```diff
@@ -178,10 +181,10 @@ function normalizeLineItem({
+    quantity
+  } = <CartItem>cartItem.node
+
+  console.log('imageURLs', cartItem)
```
This was committed by mistake.

loan-laux replied:
Is `cartItem.node` really of type `CartItem`? If so, we can leave it in.

tedraykov commented on framework/reactioncommerce/utils/normalize.ts:
```diff
@@ -136,17 +142,14 @@ export function normalizeProduct(productNode: CatalogItemProduct): Product {
-      ? groupProductOptionsByAttributeLabel(variants)
-      : [],
-    ...rest,
+    variants: !!variants ? normalizeProductVariants(<CatalogProductVariant[]>variants) : [],
```
We should define a common way for resolving the `Maybe` type because it either causes type check errors or requires redundant type castings.

loan-laux replied:
Agreed. This can be part of another PR though.

loan-laux commented on framework/reactioncommerce/utils/normalize.ts:
```diff
@@ -1,115 +1,123 @@
   images.map((image) => ({
     url: image?.URLs?.original || image?.URLs?.medium || '',
-    alt: name,
+    alt: name
```
I would usually do that too, but the trailing coma is part of Vercel's coding style. Gotta adhere to it whether we like it or not. ¯\_(ツ)_/¯

loan-laux commented on framework/reactioncommerce/utils/normalize.ts:
```diff
@@ -1,115 +1,123 @@
-    }),
+    id: variant._id,
+    displayName: variant.attributeLabel,
+    values: variant.optionTitle ? [{label: variant.optionTitle}] : []
```
Again, we need the trailing coma. I won't comment on this anymore because otherwise I would spam this review. But generally make sure that there are trailing comas in every object.

tedraykov replied:
Okay, I wasn't aware of that. Will fix it.

loan-laux reviewed (commented):
Looks good apart from the missing trailing comas in objects. Ready to merge once you take care of that!

tedraykov committed:
Merge remote-tracking branch 'origin/outgrow-reaction-commerce-provider' into outgrow-reaction-commerce-provider

tedraykov committed:
Trailing commas, null checks, formatting

Added trailing commas object literals in normalize util.
Added null checks for product and cart item normalization methods.
Formatted the code with prettier config.
Separated the nested functions in groupProductOptionsByAttributeLabel

loan-laux commented:
LGTM. Thank you very much, @tedraykov!
